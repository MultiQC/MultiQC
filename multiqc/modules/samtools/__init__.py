from .samtools import MultiqcModule

__all__ = ["MultiqcModule"]
