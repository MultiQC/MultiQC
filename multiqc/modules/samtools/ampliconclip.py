import logging
import re
from typing import Any, Dict, Optional

from multiqc import BaseMultiqcModule, config
from multiqc.plots import violin

# Initialise the logger
log = logging.getLogger(__name__)

# ampliconclip has value thing per line, documented here (search for ampliconclip):
# http://www.htslib.org/doc/samtools.html
ampliconclip_headers = {
    "ampliconclip_total_reads": {
        "title": "Total Reads",
        "description": f"Total Reads ({config.read_count_desc})",
        "shared_key": "read_count",
        "hidden": True,
        "source_col": "TOTAL READS",
    },
    "ampliconclip_total_clipped": {
        "title": "Total Clipped",
        "description": "Total Clipped ({config.read_count_desc})",
        "shared_key": "read_count",
        "hidden": True,
        "source_col": "TOTAL CLIPPED",
    },
    "ampliconclip_forward_clipped": {
        "title": "Forward Clipped",
        "description": "Forward Clipped ({config.read_count_desc})",
        "shared_key": "read_count",
        "hidden": True,
        "source_col": "FORWARD CLIPPED",
    },
    "ampliconclip_reverse_clipped": {
        "title": "Reverse Clipped",
        "description": "Reverse Clipped ({config.read_count_desc})",
        "shared_key": "read_count",
        "hidden": True,
        "source_col": "REVERSE CLIPPED",
    },
    "ampliconclip_both_clipped": {
        "title": "Both Clipped",
        "description": "Both Clipped ({config.read_count_desc})",
        "shared_key": "read_count",
        "hidden": True,
        "source_col": "BOTH CLIPPED",
    },
    "ampliconclip_not_clipped": {
        "title": "Not Clipped",
        "description": "Not Clipped ({config.read_count_desc})",
        "shared_key": "read_count",
        "hidden": True,
        "source_col": "NOT CLIPPED",
    },
    "ampliconclip_excluded_reads": {
        "title": "Excluded Reads",
        "description": "Excluded Reads({config.read_count_desc})",
        "shared_key": "read_count",
        "hidden": True,
        "source_col": "EXCLUDED",
    },
    "ampliconclip_filtered_reads": {
        "title": "Filtered Reads",
        "description": "Filtered Reads ({config.read_count_desc})",
        "shared_key": "read_count",
        "hidden": True,
        "source_col": "FILTERED",
    },
    "ampliconclip_failed_reads": {
        "title": "Failed Reads",
        "description": "Failed Reads ({config.read_count_desc})",
        "shared_key": "read_count",
        "hidden": True,
        "source_col": "FAILED",
    },
    "ampliconclip_written_reads": {
        "title": "Written Reads",
        "description": "Written Reads ({config.read_count_desc})",
        "shared_key": "read_count",
        "hidden": True,
        "source_col": "WRITTEN",
    },
}


def parse_samtools_ampliconclip(module: BaseMultiqcModule):
    """Find Samtools ampliconclip logs and parse their data"""

    samtools_ampliconclip: Dict = dict()
    for f in module.find_log_files("samtools/ampliconclip"):
        parsed_data = parse_single_report(f["f"])
        if len(parsed_data) > 0:
            if f["s_name"] in samtools_ampliconclip:
                log.debug(f"Duplicate sample name found! Overwriting: {f['s_name']}")
            module.add_data_source(f, section="ampliconclip")
            samtools_ampliconclip[f["s_name"]] = parsed_data

    # Filter to strip out ignored sample names
    samtools_ampliconclip = module.ignore_samples(samtools_ampliconclip)

    if len(samtools_ampliconclip) == 0:
        return 0

    # Superfluous function call to confirm that it is used in this module
    # Replace None with actual version if it is available
    module.add_software_version(None)

    # General Stats Table

    # Get general stats headers using the utility function, will read config.general_stats_columns
    general_stats_headers = module.get_general_stats_headers(
        all_headers={
            data_dict["source_col"]: {
                "title": data_dict["title"],
                "description": data_dict["description"],
                "shared_key": data_dict["shared_key"],
                "hidden": data_dict["hidden"],
            }
            for data_key, data_dict in ampliconclip_headers.items()
        }
    )

    # Add headers to general stats table
    if general_stats_headers:
        module.general_stats_addcols(samtools_ampliconclip, general_stats_headers, namespace="ampliconclip")

    # Make a violin plot
    keys_counts: Dict[str, Dict[str, Any]] = {
        key: {
            "shared_key": "read_count",
            "modify": None,
            "format": None,
            "suffix": None,
            "title": data["title"],
        }
        for key, data in ampliconclip_headers.items()
    }

    data_pct: Dict = dict()
    for sample, d in samtools_ampliconclip.items():
        data_pct[sample] = dict()
        total = d["ampliconclip_total_reads"]
        if total > 0:
            for metric, cnt in d.items():
                data_pct[sample][f"{metric}_pct"] = cnt / total * 100
    keys_pct = {
        f"{metric}_pct": dict(
            title=header["title"],
            min=0,
            max=100,
            suffix="%",
            shared_key=None,
        )
        for metric, header in keys_counts.items()
    }

    module.add_section(
        name="Amplicon Clip",
        anchor="samtools-ampliconclip",
        description="This module parses the output from <code>samtools ampliconclip</code>",
        plot=violin.plot(
            samtools_ampliconclip,
            headers=keys_counts,
            pconfig={
                "id": "samtools-ampliconclip-table",
                "title": "Samtools: ampliconclip: read count",
            },
        ),
    )

    module.add_section(
        name="Amplicon Clip: Percentage of total",
        anchor="samtools-ampliconclip-pct",
        description="This module parses the output from <code>samtools ampliconclip</code>",
        plot=violin.plot(
            data_pct,
            headers=keys_pct,
            pconfig={
                "id": "samtools-ampliconclip-pct-table",
                "title": "Samtools: ampliconclip: percentage of total",
            },
        ),
    )

    # Write parsed report data to a file (restructure first)
    module.write_data_file(samtools_ampliconclip, "multiqc_samtools_ampliconclip")

    # Return the number of logs that were found
    return len(samtools_ampliconclip)


def parse_single_report(file_obj):
    """
    Take a filename, parse the data assuming it's a ampliconclip file
    Returns a dictionary of metrics to value
    """
    parsed_data: Dict[str, Optional[int]] = {k: None for k in ampliconclip_headers}

    source_to_key = {data_dict["source_col"]: data_key for data_key, data_dict in ampliconclip_headers.items()}
    source_to_key["COMMAND"] = None

    for line in file_obj.splitlines():
        source, value = line.split(":", maxsplit=1)
        data_key = source_to_key[source]
        if data_key is None:
            continue
        parsed_data[data_key] = int(value)

    # check that all the values were found
    for k, v in parsed_data.items():
        if v is None:
            raise ValueError(f"Missing value '{k}' for samtools ampliconclip")

    return parsed_data
