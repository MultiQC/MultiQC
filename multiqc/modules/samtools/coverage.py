import copy

from typing import Dict, Union

import logging

from multiqc.plots import linegraph, table

log = logging.getLogger(__name__)


class CoverageReportMixin:
    def parse_samtools_coverage(self):
        """Find Samtools coverage logs and parse their data"""

        data_by_sample = dict()
        for f in self.find_log_files("samtools/coverage"):
            metrics_by_chrom = parse_single_report(f)
            if len(metrics_by_chrom) > 0:
                if f["s_name"] in data_by_sample:
                    log.debug(f"Duplicate sample name found! Overwriting: {f['s_name']}")
                self.add_data_source(f, section="coverage")
                data_by_sample[f["s_name"]] = metrics_by_chrom

        # Filter to strip out ignored sample names
        data_by_sample = self.ignore_samples(data_by_sample)
        if len(data_by_sample) == 0:
            return 0

        # Superfluous function call to confirm that it is used in this module
        # Replace None with actual version if it is available
        self.add_software_version(None)

        # Write parsed report data to a file (restructure first)
        self.write_data_file(data_by_sample, "multiqc_samtools_coverage")

        # Make a table/violin summarising the stats across all regions (mean or sum)
        self.summary_table(data_by_sample)

        # Make a line plot showing coverage stats per region, with a tab switch between stats
        self.lineplot_per_region(data_by_sample)

        # Return the number of logs that were found
        return len(data_by_sample)

    def summary_table(self, data_by_sample):
        table_data = {sname: {} for sname in data_by_sample}
        for sample, d_by_chrom in data_by_sample.items():
            vals = list(d_by_chrom.values())
            table_data[sample]["numreads"] = sum([m["numreads"] for m in vals])
            table_data[sample]["covbases"] = sum([m["covbases"] for m in vals])
            for m in vals:
                m["size"] = m["endpos"] - m["startpos"] + 1
            total_size = sum([m["size"] for m in vals])
            # Average weighted by size. Multiplying by individual weight and dividing by total weight
            table_data[sample]["coverage"] = sum([m["coverage"] * m["size"] for m in vals]) / total_size
            table_data[sample]["meandepth"] = sum([m["meandepth"] * m["size"] for m in vals]) / total_size
            table_data[sample]["meanbaseq"] = sum([m["meanbaseq"] * m["size"] for m in vals]) / total_size
            table_data[sample]["meanmapq"] = sum([m["meanmapq"] * m["size"] for m in vals]) / total_size

        headers = {
            "numreads": {
                "title": "Reads",
                "description": "Total number of mapped reads",
                "shared_key": "read_count",
                "scale": "RdYlGn",
            },
            "covbases": {
                "title": "Bases",
                "description": "Total number of mapped base pairs",
                "shared_key": "base_count",
                "scale": "Blues",
            },
            "coverage": {
                "title": "Coverage",
                "description": "Percentage of region covered with reads",
                "min": 0,
                "max": 100,
                "suffix": "%",
                "scale": "YlGn",
            },
            "meandepth": {
                "title": "Mean depth",
                "description": "Mean depth of coverage",
                "min": 0,
                "suffix": "x",
                "scale": "RdYlGn",
            },
            "meanbaseq": {
                "title": "Mean BQ",
                "description": "Mean base quality",
                "min": 0,
                "scale": "Blues",
            },
            "meanmapq": {
                "title": "Mean MQ",
                "description": "Mean mapping quality",
                "min": 0,
                "max": 60,
                "scale": "RdYlGn",
            },
        }

        self.add_section(
            name="Coverage: global stats",
            anchor="samtools-coverage-table-section",
            description=(
                "Stats parsed from <code>samtools coverage</code> output, and summarized "
                "(added up or weighted-averaged) across all regions."
            ),
            plot=table.plot(
                table_data,
                copy.deepcopy(headers),
                pconfig={
                    "id": "samtools-coverage-table",
                    "title": "Samtools Coverage: Summary",
                },
            ),
        )

        for h in headers:
            headers[h]["hidden"] = True
        headers["meandepth"]["hidden"] = False
        self.general_stats_addcols(table_data, headers, namespace="coverage")

    def lineplot_per_region(self, data_by_sample):
        tabs = {
            "numreads": {
                "title": "Mapped reads per region",
                "name": "Reads",
                "ylab": "# reads",
                "tt_suffix": " reads",
                "tt_decimals": 0,
            },
            "covbases": {
                "title": "Mapped base pairs per region",
                "name": "Bases",
                "ylab": "# bases",
                "tt_suffix": " bp",
                "tt_decimals": 0,
            },
            "coverage": {
                "title": "Percentage of region covered with reads",
                "name": "Coverage",
                "ylab": "Coverage %",
                "tt_suffix": "%",
            },
            "meandepth": {
                "title": "Mean depth per region",
                "name": "Mean depth",
                "ylab": "Depth",
                "tt_suffix": "x",
            },
            "meanbaseq": {
                "title": "Mean base quality per region",
                "name": "BQ",
                "ylab": "Base quality",
                "tt_suffix": "",
            },
            "meanmapq": {
                "title": "Mean mapping quality per region",
                "name": "MQ",
                "ylab": "Mapping quality",
                "tt_suffix": "",
            },
        }
        datasets = [
            {
                sample: {chrom: metrics[metric] for chrom, metrics in data_by_sample[sample].items()}
                for sample in data_by_sample
            }
            for metric in tabs
        ]
        data_labels = list(tabs.values())
        for dconfig in data_labels:
            dconfig["title"] = "Samtools coverage: " + dconfig["title"]

        self.add_section(
            name="Coverage: stats per region",
            description="Per-region stats parsed from <code>samtools coverage</code> output.",
            anchor="samtools-coverage-section",
            plot=linegraph.plot(
                datasets,
                {
                    "id": "samtools-coverage",
                    "title": data_labels[0]["title"],
                    "xlab": "Region",
                    "ylab": data_labels[0]["ylab"],
                    "ymin": 0,
                    "categories": True,
                    "smooth_points": 500,
                    "logswitch": True,
                    "hide_zero_cats": False,
                    "data_labels": data_labels,
                },
            ),
        )


EXPECTED_COLUMNS = [
    "rname",
    "startpos",
    "endpos",
    "numreads",
    "covbases",
    "coverage",
    "meandepth",
    "meanbaseq",
    "meanmapq",
]


def parse_single_report(f) -> Dict[str, Dict[str, Union[int, float]]]:
    """
    Example:
    #rname	startpos	endpos	numreads	covbases	coverage	meandepth	meanbaseq	meanmapq
    oligo_1512_adapters	1	156	10	156	100	7.26282	18.4	29.6
    oligo_741_adapters	1	156	0	0	0	0	0	0

    Returns a dictionary with the contig name (rname) as the key and the rest of the fields as a dictionary
    """
    parsed_data = {}
    lines = f["f"].splitlines()
    expected_header = "#" + "\t".join(EXPECTED_COLUMNS)
    if lines[0] != expected_header:
        logging.warning(f"Expected header for samtools coverage: {expected_header}, got: {lines[0]}")
        return {}

    for line in lines[1:]:
        fields = line.strip().split("\t")
        if len(fields) != len(EXPECTED_COLUMNS):
            logging.warning(f"Skipping line with {len(fields)} fields, expected {len(EXPECTED_COLUMNS)}: {line}")
        rname, startpos, endpos, numreads, covbases, coverage, meandepth, meanbaseq, meanmapq = fields
        parsed_data[rname] = dict(
            startpos=int(startpos),
            endpos=int(endpos),
            numreads=int(numreads),
            covbases=int(covbases),
            coverage=float(coverage),
            meandepth=float(meandepth),
            meanbaseq=float(meanbaseq),
            meanmapq=float(meanmapq),
        )

    return parsed_data
