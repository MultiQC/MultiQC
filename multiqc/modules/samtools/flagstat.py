# coding: utf-8
""" MultiQC submodule to parse output from Samtools flagstat """

import logging
import re
from collections import OrderedDict

from multiqc import config
from multiqc.plots import beeswarm

# Initialise the logger
log = logging.getLogger(__name__)


class FlagstatReportMixin:
    def parse_samtools_flagstats(self):
        """Find Samtools flagstat logs and parse their data"""

        self.samtools_flagstat = dict()
        for f in self.find_log_files("samtools/flagstat"):
            parsed_data = parse_single_report(f["f"])
            if len(parsed_data) > 0:
                if f["s_name"] in self.samtools_flagstat:
                    log.debug("Duplicate sample name found! Overwriting: {}".format(f["s_name"]))
                self.add_data_source(f, section="flagstat")
                self.samtools_flagstat[f["s_name"]] = parsed_data

                # Superfluous function call to confirm that it is used in this module
                # Replace None with actual version if it is available
                self.add_software_version(None, f["s_name"])

        # Filter to strip out ignored sample names
        self.samtools_flagstat = self.ignore_samples(self.samtools_flagstat)

        if len(self.samtools_flagstat) > 0:
            # Write parsed report data to a file (restructure first)
            self.write_data_file(self.samtools_flagstat, "multiqc_samtools_flagstat")

            # General Stats Table
            flagstats_headers = OrderedDict()
            flagstats_headers["flagstat_total"] = {
                "title": "{} Reads".format(config.read_count_prefix),
                "description": "Total reads in the bam file ({})".format(config.read_count_desc),
                "min": 0,
                "modify": lambda x: x * config.read_count_multiplier,
                "shared_key": "read_count",
                "hidden": True,
            }
            flagstats_headers["mapped_passed"] = {
                "title": "{} Reads Mapped".format(config.read_count_prefix),
                "description": "Reads Mapped in the bam file ({})".format(config.read_count_desc),
                "min": 0,
                "modify": lambda x: x * config.read_count_multiplier,
                "shared_key": "read_count",
            }
            flagstats_headers["mapped_passed_pct"] = {
                "title": "% Reads Mapped",
                "description": "% Reads Mapped in the bam file",
                "min": 0,
                "max": 100,
                "suffix": "%",
                "scale": "RdYlGn",
                "hidden": True,
            }
            self.general_stats_addcols(self.samtools_flagstat, flagstats_headers)

            # Make dot plot of counts
            keys = OrderedDict()
            reads = {
                "min": 0,
                "modify": lambda x: float(x) * config.read_count_multiplier,
                "suffix": "{} reads".format(config.read_count_prefix),
                "decimalPlaces": 2,
                "shared_key": "read_count",
            }
            keys["flagstat_total"] = dict(reads, title="Total Reads")
            keys["total_passed"] = dict(reads, title="Total Passed QC")
            keys["mapped_passed"] = dict(reads, title="Mapped")

            if any(v.get("secondary_passed") for v in self.samtools_flagstat.values()):
                keys["secondary_passed"] = dict(reads, title="Secondary Alignments")

            if any(v.get("supplementary_passed") for v in self.samtools_flagstat.values()):
                keys["supplementary_passed"] = dict(reads, title="Supplementary Alignments")

            keys["duplicates_passed"] = dict(reads, title="Duplicates")
            keys["paired in sequencing_passed"] = dict(reads, title="Paired in Sequencing")
            keys["properly paired_passed"] = dict(reads, title="Properly Paired")
            keys["with itself and mate mapped_passed"] = dict(
                reads, title="Self and mate mapped", description="Reads with itself and mate mapped"
            )
            keys["singletons_passed"] = dict(reads, title="Singletons")
            keys["with mate mapped to a different chr_passed"] = dict(
                reads, title="Mate mapped to diff chr", description="Mate mapped to different chromosome"
            )
            keys["with mate mapped to a different chr (mapQ >= 5)_passed"] = dict(
                reads, title="Diff chr (mapQ >= 5)", description="Mate mapped to different chromosome (mapQ >= 5)"
            )

            self.add_section(
                name="Samtools Flagstat",
                anchor="samtools-flagstat",
                description="This module parses the output from <code>samtools flagstat</code>. All numbers in millions.",
                plot=beeswarm.plot(self.samtools_flagstat, keys, {"id": "samtools-flagstat-dp"}),
            )

        # Return the number of logs that were found
        return len(self.samtools_flagstat)


#  flagstat has one thing per line, documented here (search for flagstat):
#  http://www.htslib.org/doc/samtools.html
flagstat_regexes = {
    "total": r"(\d+) \+ (\d+) in total \(QC-passed reads \+ QC-failed reads\)",
    "secondary": r"(\d+) \+ (\d+) secondary",
    "supplementary": r"(\d+) \+ (\d+) supplementary",
    "duplicates": r"(\d+) \+ (\d+) duplicates",
    "mapped": r"(\d+) \+ (\d+) mapped \((.+):(.+)\)",
    "paired in sequencing": r"(\d+) \+ (\d+) paired in sequencing",
    "read1": r"(\d+) \+ (\d+) read1",
    "read2": r"(\d+) \+ (\d+) read2",
    "properly paired": r"(\d+) \+ (\d+) properly paired \((.+):(.+)\)",
    "with itself and mate mapped": r"(\d+) \+ (\d+) with itself and mate mapped",
    "singletons": r"(\d+) \+ (\d+) singletons \((.+):(.+)\)",
    "with mate mapped to a different chr": r"(\d+) \+ (\d+) with mate mapped to a different chr",
    "with mate mapped to a different chr (mapQ >= 5)": r"(\d+) \+ (\d+) with mate mapped to a different chr \(mapQ>=5\)",
}


def parse_single_report(file_obj):
    """
    Take a filename, parse the data assuming it's a flagstat file
    Returns a dictionary {'lineName_pass' : value, 'lineName_fail' : value}
    """
    parsed_data = {}

    re_groups = ["passed", "failed", "passed_pct", "failed_pct"]
    for k, r in flagstat_regexes.items():
        r_search = re.search(r, file_obj, re.MULTILINE)
        if r_search:
            for i, j in enumerate(re_groups):
                try:
                    key = "{}_{}".format(k, j)
                    val = r_search.group(i + 1).strip("% ")
                    parsed_data[key] = float(val) if ("." in val) else int(val)
                except IndexError:
                    pass  # Not all regexes have percentages
                except ValueError:
                    parsed_data[key] = float("nan")
    # Work out the total read count
    try:
        parsed_data["flagstat_total"] = parsed_data["total_passed"] + parsed_data["total_failed"]
    except KeyError:
        pass
    return parsed_data
