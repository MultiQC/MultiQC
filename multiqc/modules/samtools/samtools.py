#!/usr/bin/env python
""" MultiQC module to parse output from Samtools """
from __future__ import print_function
from collections import OrderedDict
import logging

from multiqc import config
from multiqc.modules.base_module import BaseMultiqcModule

# Import the Samtools submodules
from .stats import StatsReportMixin
from .flagstat import FlagstatReportMixin
from .idxstats import IdxstatsReportMixin
from .rmdup import RmdupReportMixin

# Initialise the logger
log = logging.getLogger(__name__)


class MultiqcModule(BaseMultiqcModule, StatsReportMixin, FlagstatReportMixin, IdxstatsReportMixin, RmdupReportMixin):
    """ Samtools has a number of different commands and outputs.
    This MultiQC module supports some but not all. The code for
    each script is split into its own file and adds a section to
    the module output if logs are found. """

    def __init__(self):

        # Initialise the parent object
        super(MultiqcModule, self).__init__(
            name='Samtools',
            anchor='Samtools', target='Samtools',
            href='http://www.htslib.org',
            info=(" is a suite of programs for interacting with "
                  "high-throughput sequencing data."))

        # Set up class objects to hold parsed data
        self.general_stats_headers = OrderedDict()
        self.general_stats_data = dict()
        n = dict()

        # Call submodule functions
        n['stats'] = self.parse_samtools_stats()
        if n['stats'] > 0:
            log.info("Found {} stats reports".format(n['stats']))

        n['flagstat'] = self.parse_samtools_flagstats()
        if n['flagstat'] > 0:
            log.info("Found {} flagstat reports".format(n['flagstat']))

        n['idxstats'] = self.parse_samtools_idxstats()
        if n['idxstats'] > 0:
            log.info("Found {} idxstats reports".format(n['idxstats']))

        n['rmdup'] = self.parse_samtools_rmdup()
        if n['rmdup'] > 0:
            log.info("Found {} rmdup reports".format(n['rmdup']))

        # Exit if we didn't find anything
        if sum(n.values()) == 0:
            raise UserWarning

        # Add to the General Stats table (has to be called once per MultiQC module)
        self.general_stats_addcols(self.general_stats_data, self.general_stats_headers)
