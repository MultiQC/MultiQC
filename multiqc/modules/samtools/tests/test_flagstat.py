import os
from pathlib import Path

import math
import pytest

from multiqc import report
from multiqc.modules.samtools import MultiqcModule
from multiqc.utils import testing


@pytest.fixture
def data_dir():
    return testing.data_dir()


def test_data_parsed(data_dir):
    data_subdir = data_dir / "modules/samtools/flagstat"
    for path in os.listdir(data_subdir):
        path = data_subdir / path
        report.analysis_files = [path]
        report.search_files(["samtools"])
        m = MultiqcModule()
        assert len(m.saved_raw_data) > 0
        assert m.clean_s_name(Path(path).name) in list(m.saved_raw_data.values())[0]


def slurp_file(data_dir, fname):
    with (data_dir / "modules/samtools/flagstat" / fname).open() as fh:
        return fh.read()


@pytest.fixture
def rep1(data_dir):
    return slurp_file(data_dir, "small.samtools13.flagstat.log.txt")


@pytest.fixture
def rep2(data_dir):
    return slurp_file(data_dir, "small.samtools12.flagstat.log.txt")


def test_rep1(rep1):
    """Test that parsing rep1 produces expected results"""
    from multiqc.modules.samtools.flagstat import parse_single_report

    res1 = parse_single_report(rep1)

    # I expect 13 + 13 + 3 + 3 + 1 things reported in total
    assert len(res1) == 13 + 13 + 3 + 3 + 1

    assert (res1["total_passed"], res1["total_failed"]) == (5414, 0)

    assert res1["flagstat_total"] == 5414

    assert res1["mapped_passed_pct"] == 98.82

    # I expect mapped_failed_pct to be float('nan')
    assert math.isnan(res1["mapped_failed_pct"])


def test_rep2(rep1, rep2):
    """I expect rep2 to give identical results to rep1."""
    from multiqc.modules.samtools.flagstat import parse_single_report

    res1 = parse_single_report(rep1)
    res2 = parse_single_report(rep2)

    # But since nan != nan we have to strip these out.
    nans = [k for k, v in res1.items() if math.isnan(v)]
    for k in nans:
        del res1[k]
        del res2[k]

    assert res1 == res2
