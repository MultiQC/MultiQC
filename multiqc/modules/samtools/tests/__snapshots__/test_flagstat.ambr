# serializer version: 1
# name: TestSamtoolsFlagstatSnapshot.test_flagstat_snapshot
  dict({
    'multiqc_samtools_flagstat': dict({
      'low_quality_1': dict({
        'duplicates_failed': 0,
        'duplicates_passed': 0,
        'flagstat_total': 137147,
        'mapped_failed': 0,
        'mapped_failed_pct': nan,
        'mapped_passed': 137147,
        'mapped_passed_pct': 100.0,
        'paired in sequencing_failed': 0,
        'paired in sequencing_passed': 137147,
        'properly paired_failed': 0,
        'properly paired_failed_pct': nan,
        'properly paired_passed': 44296,
        'properly paired_passed_pct': 32.3,
        'read1_failed': 0,
        'read1_passed': 66552,
        'read2_failed': 0,
        'read2_passed': 70595,
        'singletons_failed': 0,
        'singletons_failed_pct': nan,
        'singletons_passed': 79413,
        'singletons_passed_pct': 57.9,
        'total_failed': 0,
        'total_passed': 137147,
        'with itself and mate mapped_failed': 0,
        'with itself and mate mapped_passed': 57734,
        'with mate mapped to a different chr (mapQ >= 5)_failed': 0,
        'with mate mapped to a different chr (mapQ >= 5)_passed': 668,
        'with mate mapped to a different chr_failed': 0,
        'with mate mapped to a different chr_passed': 14598,
      }),
      'low_quality_2': dict({
        'duplicates_failed': 0,
        'duplicates_passed': 0,
        'flagstat_total': 120672,
        'mapped_failed': 0,
        'mapped_failed_pct': nan,
        'mapped_passed': 120672,
        'mapped_passed_pct': 100.0,
        'paired in sequencing_failed': 0,
        'paired in sequencing_passed': 120672,
        'properly paired_failed': 0,
        'properly paired_failed_pct': nan,
        'properly paired_passed': 38796,
        'properly paired_passed_pct': 32.15,
        'read1_failed': 0,
        'read1_passed': 59601,
        'read2_failed': 0,
        'read2_passed': 61071,
        'singletons_failed': 0,
        'singletons_failed_pct': nan,
        'singletons_passed': 73210,
        'singletons_passed_pct': 60.67,
        'total_failed': 0,
        'total_passed': 120672,
        'with itself and mate mapped_failed': 0,
        'with itself and mate mapped_passed': 47462,
        'with mate mapped to a different chr (mapQ >= 5)_failed': 0,
        'with mate mapped to a different chr (mapQ >= 5)_passed': 508,
        'with mate mapped to a different chr_failed': 0,
        'with mate mapped to a different chr_passed': 10266,
      }),
      'samtools': dict({
        'duplicates_failed': 0,
        'duplicates_passed': 0,
        'flagstat_total': 11540659,
        'mapped_failed': 0,
        'mapped_failed_pct': nan,
        'mapped_passed': 11540643,
        'mapped_passed_pct': 100.0,
        'paired in sequencing_failed': 0,
        'paired in sequencing_passed': 11534294,
        'properly paired_failed': 0,
        'properly paired_failed_pct': nan,
        'properly paired_passed': 11518384,
        'properly paired_passed_pct': 99.86,
        'read1_failed': 0,
        'read1_passed': 5767147,
        'read2_failed': 0,
        'read2_passed': 5767147,
        'secondary_failed': 0,
        'secondary_passed': 6365,
        'singletons_failed': 0,
        'singletons_failed_pct': nan,
        'singletons_passed': 12,
        'singletons_passed_pct': 0.0,
        'supplementary_failed': 0,
        'supplementary_passed': 0,
        'total_failed': 0,
        'total_passed': 11540659,
        'with itself and mate mapped_failed': 0,
        'with itself and mate mapped_passed': 11534266,
        'with mate mapped to a different chr (mapQ >= 5)_failed': 0,
        'with mate mapped to a different chr (mapQ >= 5)_passed': 344,
        'with mate mapped to a different chr_failed': 0,
        'with mate mapped to a different chr_passed': 628,
      }),
      'small': dict({
        'duplicates_failed': 0,
        'duplicates_passed': 0,
        'flagstat_total': 5414,
        'mapped_failed': 0,
        'mapped_failed_pct': nan,
        'mapped_passed': 5350,
        'mapped_passed_pct': 98.82,
        'paired in sequencing_failed': 0,
        'paired in sequencing_passed': 5401,
        'properly paired_failed': 0,
        'properly paired_failed_pct': nan,
        'properly paired_passed': 5011,
        'properly paired_passed_pct': 92.78,
        'read1_failed': 0,
        'read1_passed': 2709,
        'read2_failed': 0,
        'read2_passed': 2692,
        'secondary_failed': 0,
        'secondary_passed': 13,
        'singletons_failed': 0,
        'singletons_failed_pct': nan,
        'singletons_passed': 64,
        'singletons_passed_pct': 1.18,
        'supplementary_failed': 0,
        'supplementary_passed': 0,
        'total_failed': 0,
        'total_passed': 5414,
        'with itself and mate mapped_failed': 0,
        'with itself and mate mapped_passed': 5273,
        'with mate mapped to a different chr (mapQ >= 5)_failed': 0,
        'with mate mapped to a different chr (mapQ >= 5)_passed': 81,
        'with mate mapped to a different chr_failed': 0,
        'with mate mapped to a different chr_passed': 206,
      }),
    }),
  })
# ---
# name: test_individual_flagstat_file[small.samtools12.flagstat.log.txt]
  dict({
    'multiqc_samtools_flagstat': dict({
      'small': dict({
        'duplicates_failed': 0,
        'duplicates_passed': 0,
        'flagstat_total': 5414,
        'mapped_failed': 0,
        'mapped_failed_pct': nan,
        'mapped_passed': 5350,
        'mapped_passed_pct': 98.82,
        'paired in sequencing_failed': 0,
        'paired in sequencing_passed': 5401,
        'properly paired_failed': 0,
        'properly paired_failed_pct': nan,
        'properly paired_passed': 5011,
        'properly paired_passed_pct': 92.78,
        'read1_failed': 0,
        'read1_passed': 2709,
        'read2_failed': 0,
        'read2_passed': 2692,
        'secondary_failed': 0,
        'secondary_passed': 13,
        'singletons_failed': 0,
        'singletons_failed_pct': nan,
        'singletons_passed': 64,
        'singletons_passed_pct': 1.18,
        'supplementary_failed': 0,
        'supplementary_passed': 0,
        'total_failed': 0,
        'total_passed': 5414,
        'with itself and mate mapped_failed': 0,
        'with itself and mate mapped_passed': 5273,
        'with mate mapped to a different chr (mapQ >= 5)_failed': 0,
        'with mate mapped to a different chr (mapQ >= 5)_passed': 81,
        'with mate mapped to a different chr_failed': 0,
        'with mate mapped to a different chr_passed': 206,
      }),
    }),
  })
# ---
# name: test_individual_flagstat_file[small.samtools13.flagstat.log.txt]
  dict({
    'multiqc_samtools_flagstat': dict({
      'small': dict({
        'duplicates_failed': 0,
        'duplicates_passed': 0,
        'flagstat_total': 5414,
        'mapped_failed': 0,
        'mapped_failed_pct': nan,
        'mapped_passed': 5350,
        'mapped_passed_pct': 98.82,
        'paired in sequencing_failed': 0,
        'paired in sequencing_passed': 5401,
        'properly paired_failed': 0,
        'properly paired_failed_pct': nan,
        'properly paired_passed': 5011,
        'properly paired_passed_pct': 92.78,
        'read1_failed': 0,
        'read1_passed': 2709,
        'read2_failed': 0,
        'read2_passed': 2692,
        'secondary_failed': 0,
        'secondary_passed': 13,
        'singletons_failed': 0,
        'singletons_failed_pct': nan,
        'singletons_passed': 64,
        'singletons_passed_pct': 1.18,
        'supplementary_failed': 0,
        'supplementary_passed': 0,
        'total_failed': 0,
        'total_passed': 5414,
        'with itself and mate mapped_failed': 0,
        'with itself and mate mapped_passed': 5273,
        'with mate mapped to a different chr (mapQ >= 5)_failed': 0,
        'with mate mapped to a different chr (mapQ >= 5)_passed': 81,
        'with mate mapped to a different chr_failed': 0,
        'with mate mapped to a different chr_passed': 206,
      }),
    }),
  })
# ---
